[UI] Volume Batch Creation

ChengYanJin opened the pull request:
**Component**: ui, volumes 

**Context**: 
We want to be able to batch create volume from UI.

**Summary**:
Note that:
For `RawBlockDevice`, the recommended path only happens with `'/dev/vda'`, we will see later if 
For `SparseLoopDevice`, the size of the batch volume are the same as the default one.

**Acceptance criteria**: 
1. The Default Volume Creation Form
![image](https://user-images.githubusercontent.com/18453133/102068607-ecdea800-3dfc-11eb-98fe-194b81960a6e.png)

2. Volume type `RawBlockDevice` batch creation 
![image](https://user-images.githubusercontent.com/18453133/102068646-fe27b480-3dfc-11eb-9d9e-e6a3da81dff4.png)

3. Volume type `SparseLoopDevice` batch creation. (From big HD screen)
![image](https://user-images.githubusercontent.com/18453133/102069552-37acef80-3dfe-11eb-9469-f734ab86eab8.png)


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2964 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/styles: Remove the breadcrumb in Create New Volume page

We KILL breadcrumb!

Refs: #2964

ChengYanJin committed:
ui/volumes: Set up the form of batch volume creation

Refs: #2964

ChengYanJin committed:
ui/utils: Add utility function to generate the linux drive name

Refs: #2964

ChengYanJin committed:
ui/volumes: Add dependents field base on the base field

ChengYanJin committed:
ui/package: Upgrade the version of formik to v2

Refs: #2964

ChengYanJin committed:
ui/ducks/volumes: Change the payload of createVolume to an array

Refs: #2964

ChengYanJin committed:
ui/utils: Add util function to format the data for batch volume

Refs: #2964

ChengYanJin committed:
ui/volumes: Adapt the creationVolume container for batch volume creation

Refs: #2964

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

JBWatenbergScality commented on ui/src/services/utils.js:
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
+export const linuxDrivesNamingIncrement = (string, increment) => {
+  if (string.match(/^\/dev\/vd[a-z]/)) {
+    while (increment--) {
+      var lastChar = string.slice(-1);
```
I'm not yet sure about your styleguide conventions but I would always prefer usage of let or const for variable definitions, using const here will reflect clearly that this variable doesn't change elsewhere and will slightly simplify reading this code

JBWatenbergScality commented on ui/src/services/utils.js:
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
+ *
+ * const nextDevicePath = linuxDrivesNamingIncrement(string)
+ */
+export const linuxDrivesNamingIncrement = (string, increment) => {
```
I would rename `string` to something more explicit such as `drivePath`

JBWatenbergScality replied:
Additionally for utils function like this one it seems important for me to add a unit test that covers the edge cases you want to support, but not sure if it is your traditional way of working so we can discuss about it.

ChengYanJin replied:
Indeed, it's a very bad name. I will rename it to `devicePath`.

JBWatenbergScality commented on ui/src/services/utils.js:
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
+ * @example
+ * const string = '/dev/vda'
+ *
+ * const nextDevicePath = linuxDrivesNamingIncrement(string)
```
This example lack increment parameter

JBWatenbergScality commented on ui/src/services/utils.js:
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
 };
+
+/**
+ * This function increments the string by 1
```
This description doesn't seems to be accurate

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -237,16 +241,55 @@ const CreateVolume = (props) => {
+        {!!meta.touched && !!meta.error && <div>{meta.error}</div>}
+      </>
+    );
   };
```
As those two components only differs in the way they setFieldValue, they might be factorized.

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -362,6 +413,34 @@ const CreateVolume = (props) => {
+                const preVolNum = values.numberOfVolumes;
+
+                setFieldValue(field, inputVolNum);
+                var diff = preVolNum - inputVolNum;
```
I would use const instead to declare this variable

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -362,6 +413,34 @@ const CreateVolume = (props) => {
+                var diff = preVolNum - inputVolNum;
+                if (diff > 0) {
+                  // REMOVE volume object from `values.volumes` base on the index
+                  for (var i = inputVolNum; i < preVolNum; i++) {
```
I would prefer use let to avoid any hoisting issues

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -362,6 +413,34 @@ const CreateVolume = (props) => {
+                  }
+                } else if (diff < 0) {
+                  // PUSH new volume object to `values.volumes`
+                  var absDiff = Math.abs(diff);
```
I would use const instead to declare this variable

JBWatenbergScality commented on ui/src/ducks/app/volumes.js:
```diff
@@ -118,8 +118,8 @@ export const updateStorageClassAction = (payload) => {
 
-export const createVolumeAction = (newVolume) => {
-  return { type: CREATE_VOLUMES, payload: { newVolume } };
+export const createVolumeAction = (newVolumes) => {
```
I would rename this to `createVolumesAction` to reflect the batch creation possibility

JBWatenbergScality commented on ui/src/ducks/app/volumes.js:
```diff
@@ -205,97 +205,97 @@ export function* fetchStorageClass() {
-          message: intl.translate('volume_creation_failed', {
-            name: newVolume.name,
-          }),
+          title: 'Volume Form Error',
```
Translation seems to be missing here

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -54,6 +55,8 @@ const ActionContainer = styled.div`
 const CreateVolumeLayout = styled.div`
   display: inline-block;
   margin-top: ${padding.base};
+  overflow-y: scroll;
```
Is it intended to display scrollbars by default even for non scrollable content ? You might prefer to use `auto` instead

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -622,7 +640,22 @@ const CreateVolume = (props) => {
+                              placement="right"
+                              overlay={
+                                <div style={{ minWidth: '200px' }}>
+                                  A devicePath or a partition mount point.
```
I think there is a missing translation here

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -290,9 +382,18 @@ const CreateVolume = (props) => {
-              createVolume(newVolume);
+              const newVolumes = { ...values };
+              newVolumes.size = `${values.sizeInput}${values.selectedUnit}`;
+              if (newVolumes.multiVolumeCreation) {
```
This if could be transferred in formatDataForBatchVolumeCreation as it already partially exists there and just missing the else branch to return a one element array

JBWatenbergScality commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -276,3 +276,28 @@ export const getVolumeListData = createSelector(
   },
 );
+
+export const formatDataForBatchVolumeCreation = (newVolumes) => {
```
I would add a simple unit test for the two branches of this function, also as mentioned in the usage of this function I would handle the `!multiVolumeCreation` case here and return a single entry array containing newVolumes, that would allow a simpler usage of the function as its ouput will always be of the same shape `(newVolumes: VolumeCreationFormState) => Volume[]` instead of `(newVolumes: VolumeCreationFormState) => Volume[] | void`

bert-e commented:
# Conflict




There is a conflict between your branch `feature/2964-volume-batch-creation` and the
destination branch `development/2.7`.

Please resolve the conflict on **the feature branch** (`feature/2964-volume-batch-creation`).

```bash
 $ git fetch
 $ git checkout origin/feature/2964-volume-batch-creation
 $ git merge origin/development/2.7
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/2964-volume-batch-creation
```

ChengYanJin committed:
ui/styles: Fix the scroll bar issue by overflow:auto

"overflow-y: scroll" protentially would display the scrollbar by default
even for non scrollable content, which casue the issue of we have the
scrollbar everywhere.

Refs: #2964

ChengYanJin committed:
ui/volumes: Add unit test for formatVolumeCreationData function

Handle the single volume creation case in the utility function instead
of in the createVolume component so that the return of this function
is always an array.

For single volume, return an array with the origin volume.
For batch volume, return an array with the formatted volumes.

Refs: #2964

ChengYanJin committed:
ui/volumes: Factorize the recommendField

Correct the variable declaration too

Refs: #2964

ChengYanJin committed:
ui/translation: Add missing translation for batch volume creation

Refs: #2964

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

thomasdanan reviewed (dismissed):
@ChengYanJin  I like the behavior.
I would change one little thing:
When you create multiple Volumes, the Device Path for each Volume is computed from the Global Device Path field. I think the first Volume should get what you input in the Global Device Path field (i.e. not increment the Device Path of the first Volume).

Also when playing with multiple Volume check box and the number of Volumes, you will observe strange behavior. Maybe some state not properly managed

I wonder if we should disable Global Name and Device Path global fields when creating multiple Volumes because if you override Name or Device Path and then you change the Global field, you will lose all your customizations (@JBWatenbergScality what do you think?)

Also for number increments, I would do 01, 02, 03 etc ... instead of 1, 2, 3 .... 

Thanks and well done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @thomasdanan

JBWatenbergScality commented:
@thomasdanan @ChengYanJin From my point of view once the name or path is touched we shouldn't change automatically their values to maintain the user intentions but should still let the ability for the user to change the first volume name and path. 

Also the ui might be a little bit misleading as the user is unsure if the default values will actually create a volume. To be precise mentioning `/dev/vda` by default and then creating one more volume will pre-complete the first entry with `/dev/vdb` in the current behavior but when submitting the form just `/dev/vdb` will be created.

The simplest interaction I could think of closed enough to the current implementation would be to remove the default name and path fields when checking `Create multiple volumes?` and compute the names of the volumes from the first one if subsequent one are not touched.

ChengYanJin committed:
ui/volumes: Fix the bug of input number doesn't sync with form

Ref: #2964

ChengYanJin committed:
ui/volumes: Fix the defaults path start from the input path

The recommend default device path should start from the global one

Refs: #2964

ChengYanJin committed:
ui/volumes: Change the behaviour of the default name/path

We only recommend the name/path for the empty field, means if the global
name/page get updated, the defaults won't change.

Refs: #2964

ChengYanJin committed:
ui/volumes: Improve the name of the batch volume

Refs: #2964

JBWatenbergScality commented on ui/src/services/utils.js:
```diff
@@ -479,3 +479,23 @@ export const linuxDrivesNamingIncrement = (devicePath, increment) => {
+
+/**
+ * This function formats the name base on the index
+ * @param {string} name - The name the default volume name
```
Remove redundant name `The name the default volume`

JBWatenbergScality commented on ui/src/services/utils.js:
```diff
@@ -479,3 +479,23 @@ export const linuxDrivesNamingIncrement = (devicePath, increment) => {
+ * @example
+ * const name = 'volume-test'
+ *
+ * const nextDevicePath = linuxDrivesNamingIncrement(name, 1)
```
`const formatedName = formatBatchName(name, 1) `

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -513,12 +633,108 @@ const CreateVolume = (props) => {
+                                {intl.translate('number_volume_create')}
+                              </span>
+                              {/* TODO: Generalize the number input component to core-ui. */}
+                              <InputNumberComponentStyle
```
I think we should handle a max number of volumes to avoid dom freeze in case of the user mistakenly ask 10000 volumes, simply add a max prop here should be enough

ChengYanJin replied:
Indeed, good point!
I am going to set the `max` number to`70` for the moment.

ChengYanJin committed:
ui/volumes: Fix the default path bug when the index is over 53

Set the max number of the input volume creation to 70, since it frozes
the UI when rendering the form.
We may even want to lower the max number depents on the use case.

Refs: #2964

ChengYanJin committed:
ui/volumes: Set the max number of volume batch creation to 70

Refs: #2964

ChengYanJin committed:
ui/volumes: When re-check the box only update the untouched fields

In order to keep the user customization, edit the global value
(name/path) should not affect the defaults pre-fill of the touched
field.

Unless the users uncheck the box of multi-volume creation and re-check
again, we only update the untouched field with the new global value.

Refs: #2964

ChengYanJin commented:
Hello @thomasdanan @JBWatenbergScality ,

According to the feedback from you and our offline discussions. I've addressed the following issues. 

1.
> Thomas: When you create multiple Volumes, the Device Path for each Volume is computed from the Global Device Path field. I think the first Volume should get what you input in the Global Device Path field (i.e. not increment the Device Path of the first Volume).

Fix the bug of the Device path, the first volume should be the same as the Global field.

2.
> Thomas: Also when playing with multiple Volume check box and the number of Volumes, you will observe strange behavior. Maybe some state not properly managed

Fix the bug, will add interaction test of the form in the following PR.

3.
> Thomas: I wonder if we should disable Global Name and Device Path global fields when creating multiple Volumes because if you override Name or Device Path and then you change the Global field, you will lose all your customizations (@JBWatenbergScality what do you think?)

 In order to keep the user customization, edit the global value(name/path) should not affect the defaults pre-fill of the touched field.
Unless the users uncheck the box of multi-volume creation and re-check again, we only update the untouched field with the new global value.

4. 
> Thomas: Also for number increments, I would do 01, 02, 03 etc ... instead of 1, 2, 3 ....

Done.

5. Set the maximum number of the volume to create `70` to prevent the UI frozen. 
The UI should pop up the error message when the input number is bigger than 70 (Waiting for the [PR #213](https://github.com/scality/core-ui/pull/213) in core-ui to be landed)

Please let me know if you see any improvements/ bugs to be done. 
Thank you!

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -421,6 +421,33 @@ const CreateVolume = (props) => {
+                setFieldValue(field, selectedObj.target.checked);
+                // Clear the untouched field to get the update from the global value
+                for (let i = 0; i < values.numberOfVolumes; i++) {
+                  if (
+                    touched &&
+                    touched.volumes &&
+                    touched?.volumes[i]?.name === true
+                  ) {
+                    continue;
+                  } else {
+                    setFieldValue(`volumes[${i}]name`, '');
+                  }
+                  if (
+                    touched &&
+                    touched.volumes &&
+                    touched?.volumes[i]?.path === true
+                  ) {
+                    continue;
+                  } else {
+                    setFieldValue(`volumes[${i}]path`, '');
+                  }
```
I think this can be simplified to : 
```js
if (!touched?.volumes[i]?.name) {
    setFieldValue(`volumes[${i}]name`, '');
}
if (!touched?.volumes[i]?.path) {
    setFieldValue(`volumes[${i}]path`, '');
}
```

ChengYanJin replied:
Hi @JBWatenbergScality ,

Yes. Good point. But I guess we need to use optional chaining to make sure the Volumes array exists.
Please see:

```js
                  if (!touched?.volumes?.[i]?.name) {
                    setFieldValue(`volumes[${i}]name`, '');
                  }
                  if (!touched?.volumes?.[i]?.path) {
                    setFieldValue(`volumes[${i}]path`, '');
                  }
```

ChengYanJin commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @thomasdanan







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye chengyanjin.
